src/ucs/type/spinlock.h

```diff
@@ -40,50 +51,70 @@ typedef struct ucs_recursive_spinlock {
 } ucs_recursive_spinlock_t;
 
 
-static ucs_status_t ucs_spinlock_init(ucs_spinlock_t *lock, int flags)
+void ucs_recursive_spinlock_destroy(ucs_recursive_spinlock_t *lock);
+
+/**
+ * Spinlock implementation section
+ */
+static UCS_F_ALWAYS_INLINE void ucs_spinlock_init(ucs_spinlock_t *lock)
+{
+    lock->lock = UCS_SPINLOCK_FREE;
+}
+
+static UCS_F_ALWAYS_INLINE void ucs_spinlock_destroy(ucs_spinlock_t *lock)
 {
-    int ret, lock_flags;
+}
 
-    if (flags & UCS_SPINLOCK_FLAG_SHARED) {
-        lock_flags = PTHREAD_PROCESS_SHARED;
-    } else {
-        lock_flags = PTHREAD_PROCESS_PRIVATE;
-    }
+static UCS_F_ALWAYS_INLINE int ucs_spin_try_lock(ucs_spinlock_t *lock)
+{
+    return ucs_atomic_cswap32(&lock->lock, UCS_SPINLOCK_FREE,
+                              UCS_SPINLOCK_BUSY) == UCS_SPINLOCK_FREE;
+}
 
-    ret = pthread_spin_init(&lock->lock, lock_flags);
-    if (ret != 0) {
-        return UCS_ERR_IO_ERROR;
+static UCS_F_ALWAYS_INLINE void ucs_spin_lock(ucs_spinlock_t *lock)
+{
+    while (!ucs_spin_try_lock(lock)) {
+        while (lock->lock != UCS_SPINLOCK_FREE) {
+            /* spin */
+#if defined(__x86_64__)
```

Comment:
Probably should go to arch and avoid arch ifdef here

Reply:
More I think about this more I get convinced the entire has to go to arch: https://elixir.bootlin.com/linux/v4.4/source/arch/arm64/include/asm/spinlock.h

Reply:
cpu/arch is not installed as external header and we have to use arch macro. yes we can add more arch-specific infra, but it will be overcomplicated for this case.

here used memory fence read to spin. could you explain why this solution is not acceptable?

Reply:
updated ARM implementation of spin-wait using wife/sevl

Reply:
@shamisp unfortunately we cannot install cpu.h because it depends on config.h
it's needed because of #939 , that defined UCS_SYS_CACHE_LINE_SIZE in configure time